runtime/oti/VMHelpers.hpp

```diff
@@ -998,6 +998,190 @@ class VM_VMHelpers
 		return consumed;
 	}
 
+	/**
+	 * Computes the hash value for an input string using the hash algorithm defined by the java/lang/String.hashCode()I
+	 * method.
+	 *
+	 * @param data points to raw UTF8 bytes, assumed to be a valid (potentially multi-byte) encoding
+	 * @param length the number of bytes to hash
+	 *
+	 * @return hash code for the UTF8 string
+	 */
+	static VMINLINE UDATA
+	computeHashForUTF8(const U_8 *data, UDATA length)
+	{
+		UDATA hash = 0;
+		const U_8 * end = data + length;
+
+		while (data < end) {
+			U_16 c;
+
+			data += decodeUTF8Char(data, &c);
+			hash = (hash << 5) - hash + c;
+		}
+		return hash;
+	}
+
+	/**
+	 * Computes the hash value for an input string using the hash algorithm defined by the java/lang/String.hashCode()I
+	 * method.
+	 *
+	 * @param data points to raw UTF8 bytes, all of which are within the ASCII subset ord. [0, 127]
+	 * @param length the number of bytes to hash
+	 *
+	 * @return hash code for the UTF8 string
+	 */
+	static VMINLINE UDATA
+	computeHashForASCII(const U_8 *data, UDATA length)
+	{
+		UDATA hash = 0;
+		for (UDATA i = 0; i < length; ++i) {
+			hash = (hash << 5) - hash + data[i];
+		}
+		return hash;
+	}
+	
+	/**
+	 * Determines whether the UTF8 string is an ASCII string.
+	 *
+	 * @param data[in] points to raw UTF8 bytes
+	 * @param length[in] the number of bytes representing characters in data
+	 *
+	 * @return true if all of the characters in the UTF8 input string are ASCII; false otherwise
+	 */
+	static VMINLINE bool
+	isUTF8ASCII(const U_8 *data, UDATA length)
+	{
+		bool isASCII = true;
+
+		for (UDATA i = 0; i < length; ++i) {
+			if (data[i] > 0x7F) {
+				isASCII = false;
+				break;
+			}
+		}
+
+		return isASCII;
+	}
+
+	/**
+	 * Copies a UTF8 string into a UTF16 string at a specific index.
```

Comment:
Please add a comment that indicates this and `copyUTF8ToASCII` have to be kept in sync.  Both functions need the comment added to their documentation

Reply:
Fixed in c86ff40